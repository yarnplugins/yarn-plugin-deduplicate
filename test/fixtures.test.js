const childProcess = require("child_process");
const { promises: fs, readdirSync } = require("fs");
const os = require("os");
const path = require("path");
const rimraf = require("rimraf");

describe("fixtures", () => {
	const fixtureDirPath = path.resolve(__dirname, "./fixtures/");

	test.each(readdirSync(fixtureDirPath))("fixture %s", async fixtureName => {
		const fixturePath = path.join(fixtureDirPath, fixtureName);
		const tmpdir = path.join(
			os.tmpdir(),
			`yarn-deduplicate-test-${fixtureName}`
		);
		console.log(tmpdir)

		// prepare
		rimraf.sync(tmpdir);
		await fs.mkdir(tmpdir);
		await copyDir(fixturePath, tmpdir);
		await fs.copyFile(
			path.resolve(__dirname, "../.yarn/releases/yarn-2.x.cjs"),
			path.join(tmpdir, "yarn.js")
		);
		await fs.copyFile(
			path.resolve(
				__dirname,
				"../bundles/@yarnpkg/plugin-deduplicate.js"
			),
			path.join(tmpdir, "yarn-deduplicate.js")
		);
		await fs.writeFile(
			path.join(tmpdir, ".yarnrc.yml"),
			`
enableTimers: false
plugins:
  - yarn-deduplicate.js

yarnPath: yarn.js`,
			{ encoding: "utf8" }
		);
		childProcess.execSync(`git init`, { cwd: tmpdir });
		childProcess.execSync(`git config user.name "jest"`, { cwd: tmpdir });
		childProcess.execSync(`git config user.email "jest@example.com"`, {
			cwd: tmpdir
		});
		childProcess.execSync(`git add -A`, { cwd: tmpdir });
		childProcess.execSync(`git commit -m 'Initial commit'`, { cwd: tmpdir });

		const { stdout, stderr } = childProcess.spawnSync(`yarn`, [`deduplicate`], {
			cwd: tmpdir,
			// env: {}
		});
		const diff = childProcess
			.execSync(`git diff --patch`, { cwd: tmpdir })
			.toString("utf8");

		expect(diff).toMatchSnapshot();
		expect((await stdout).toString("utf8")).toMatchSnapshot();
		expect((await stderr).toString("utf8")).toMatchSnapshot();
	});
});

async function copyDir(source, destination) {
	const files = await fs.readdir(source);
	return Promise.all(
		files
			.map(fileName => {
				return path.join(source, fileName);
			})
			.map(async filePath => {
				const stats = await fs.stat(filePath);
				if (stats.isDirectory()) {
					await fs.mkdir(path.join(destination, path.basename(filePath)));
					return copyDir(
						filePath,
						path.join(destination, path.basename(filePath))
					);
				}

				return fs.copyFile(
					filePath,
					path.join(destination, path.basename(filePath))
				);
			})
	);
}

async function folderMatches(left, right) {
	// ignore extraneous files from right (these are cli/plugin binaries and the cache)
	const files = await fs.readdir(left);
}
