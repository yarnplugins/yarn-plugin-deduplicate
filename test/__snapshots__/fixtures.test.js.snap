// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixtures fixture test-a 1`] = `
"diff --git a/.pnp.js b/.pnp.js
index 9c1636d..b9d7afd 100755
--- a/.pnp.js
+++ b/.pnp.js
@@ -51,19 +51,12 @@ function $$SETUP_STATE(hydrateRuntimeState) {
         [null, {
           \\"packageLocation\\": \\"./\\",
           \\"packageDependencies\\": [
-            [\\"@babel/runtime\\", \\"npm:7.5.0\\"],
+            [\\"@babel/runtime\\", \\"npm:7.6.0\\"],
             [\\"@testing-library/dom\\", \\"npm:6.2.0\\"]
           ]
         }]
       ]],
       [\\"@babel/runtime\\", [
-        [\\"npm:7.5.0\\", {
-          \\"packageLocation\\": \\"./.yarn/cache/@babel-runtime-npm-7.5.0-7f6c1aead4.zip/node_modules/@babel/runtime/\\",
-          \\"packageDependencies\\": [
-            [\\"@babel/runtime\\", \\"npm:7.5.0\\"],
-            [\\"regenerator-runtime\\", \\"npm:0.13.3\\"]
-          ]
-        }],
         [\\"npm:7.6.0\\", {
           \\"packageLocation\\": \\"./.yarn/cache/@babel-runtime-npm-7.6.0-b4d272f528.zip/node_modules/@babel/runtime/\\",
           \\"packageDependencies\\": [
@@ -198,7 +191,7 @@ function $$SETUP_STATE(hydrateRuntimeState) {
           \\"packageLocation\\": \\"./\\",
           \\"packageDependencies\\": [
             [\\"before\\", \\"workspace:.\\"],
-            [\\"@babel/runtime\\", \\"npm:7.5.0\\"],
+            [\\"@babel/runtime\\", \\"npm:7.6.0\\"],
             [\\"@testing-library/dom\\", \\"npm:6.2.0\\"]
           ]
         }]
diff --git a/.yarn/cache/@babel-runtime-npm-7.5.0-7f6c1aead4.zip b/.yarn/cache/@babel-runtime-npm-7.5.0-7f6c1aead4.zip
deleted file mode 100644
index 075037b..0000000
Binary files a/.yarn/cache/@babel-runtime-npm-7.5.0-7f6c1aead4.zip and /dev/null differ
diff --git a/yarn.lock b/yarn.lock
index 9b3bc72..358b9cb 100644
--- a/yarn.lock
+++ b/yarn.lock
@@ -4,16 +4,7 @@
 __metadata:
   version: 3
 
-\\"@babel/runtime@npm:^7.5.0\\":
-  version: 7.5.0
-  resolution: \\"@babel/runtime@npm:7.5.0\\"
-  dependencies:
-    regenerator-runtime: \\"npm:^0.13.2\\"
-  checksum: c853ae00d007fc1f5bb0cf1de80abebf0cade41b1eff8e81641b5f145493c986b4e31d6f7c3485c4cb890212373d7145481e033a0aad406e09027cc6df3075f9
-  languageName: node
-  linkType: hard
-
-\\"@babel/runtime@npm:^7.5.5\\":
+\\"@babel/runtime@npm:^7.5.0, @babel/runtime@npm:^7.5.5\\":
   version: 7.6.0
   resolution: \\"@babel/runtime@npm:7.6.0\\"
   dependencies:
"
`;

exports[`fixtures fixture test-a 2`] = `
"➤ YN0000: ┌ deduplication step
➤ YN0000: │ @babel/runtime@npm:^7.5.0 can be deduplicated from runtime@7.5.0 to runtime@7.6.0
➤ YN0000: └ Completed
➤ YN0000: ┌ Resolution step
➤ YN0000: └ Completed
➤ YN0000: ┌ Fetch step
➤ YN0019: │ @babel-runtime-npm-7.5.0-7f6c1aead4.zip appears to be unused - removing
➤ YN0000: └ Completed
➤ YN0000: ┌ Link step
➤ YN0000: └ Completed
➤ YN0000: Done
"
`;
